x/sanction/keeper/keeper.go

```diff
@@ -0,0 +1,408 @@
+		if cb(addr) {
+			break
+		}
+	}
+}
+
+// getTemporaryEntryPrefixStore returns a kv store prefixed for temporary sanction/unsanction entries, and the prefix bytes used.
+// If an addr is provided, the store is prefixed for just the given address.
+// If addr is empty, it will be prefixed for all temporary entries.
+func (k Keeper) getTemporaryEntryPrefixStore(ctx sdk.Context, addr sdk.AccAddress) (sdk.KVStore, []byte) {
+	pre := CreateTemporaryAddrPrefix(addr)
+	return prefix.NewStore(ctx.KVStore(k.storeKey), pre), pre
+}
+
+// IterateTemporaryEntries iterates over each of the temporary entries.
+// If an address is provided, only the temporary entries for that address are iterated,
+// otherwise all entries are iterated.
+// The callback takes in the address in question, the governance proposal associated with it, and whether it's a sanction (true) or unsanction (false).
+// The callback should return whether to stop iteration (true = stop, false = keep going).
+func (k Keeper) IterateTemporaryEntries(ctx sdk.Context, addr sdk.AccAddress, cb func(addr sdk.AccAddress, govPropID uint64, isSanction bool) (stop bool)) {
+	store, pre := k.getTemporaryEntryPrefixStore(ctx, addr)
+
+	iter := store.Iterator(nil, nil)
+	defer iter.Close()
+
+	for ; iter.Valid(); iter.Next() {
+		key := ConcatBz(pre, iter.Key())
+		kAddr, govPropID := ParseTemporaryKey(key)
+		isSanction := IsSanctionBz(iter.Value())
+		if cb(kAddr, govPropID, isSanction) {
+			break
+		}
+	}
+}
+
+// getProposalIndexPrefixStore returns a kv store prefixed for the gov prop -> temporary sanction/unsanction index entries,
+// and the prefix bytes used.
+// If a gov prop id is provided, the store is prefixed for just that proposal.
+// If not provided, it will be prefixed for all temp index entries.
+func (k Keeper) getProposalIndexPrefixStore(ctx sdk.Context, govPropID *uint64) (sdk.KVStore, []byte) {
+	pre := CreateProposalTempIndexPrefix(govPropID)
+	return prefix.NewStore(ctx.KVStore(k.storeKey), pre), pre
+}
+
+// IterateProposalIndexEntries iterates over all of the index entries for temp entries.
+// The callback takes in the gov prop id and address.
+// The callback should return whether to stop iteration (true = stop, false = keep going).
+func (k Keeper) IterateProposalIndexEntries(ctx sdk.Context, govPropID *uint64, cb func(govPropID uint64, addr sdk.AccAddress) (stop bool)) {
+	store, pre := k.getProposalIndexPrefixStore(ctx, govPropID)
+
+	iter := store.Iterator(nil, nil)
+	defer iter.Close()
+
+	for ; iter.Valid(); iter.Next() {
+		key := ConcatBz(pre, iter.Key())
+		kPropID, kAddr := ParseProposalTempIndexKey(key)
+		if cb(kPropID, kAddr) {
+			break
+		}
+	}
+}
+
+// IsAddrThatCannotBeSanctioned returns true if the provided address is one of the ones that cannot be sanctioned.
+// Returns false if the addr can be sanctioned.
+func (k Keeper) IsAddrThatCannotBeSanctioned(addr sdk.AccAddress) bool {
+	// Okay. I know this is a clunky name for this function.
```

Comment:
NIT: I like the name and don't find it clunky. It clearly explains exactly what it does with.  No need to add more to your comments here. imo

Reply:
I originally had it named `IsUnsanctionableAddr` but kept getting it confused with `IsSanctionedAddr`, so I renamed it to `IsAddrThatCannotBeSanctioned`. I included this comment to dissuade someone else from later changing it back to `IsUnsanctionableAddr`, not realizing the confusion it can present. Primarily, I found that it was really easy to accidentally call `IsUnsanctionableAddr` instead of `IsSanctionedAddr`, and not see the mistake when reviewing code.